### PR TITLE
Add badge for the [HSTS] preload list

### LINKED ISSUE #2913: HSTS preload list status (HTTP security)
:clipboard: **Description**

A clear and concise description of the new badge.

- Which service is this badge for
  - [Chrome's HTTP Strict Transport Security (HSTS) preload list](https://en.wikipedia.org/wiki/HTTP_Strict_Transport_Security)
- What sort of information should this badge show?
  - HSTS status `hsts | status`
  - Status may be `preloaded`, `pending` or `unknow`
  - Link should points to https://hstspreload.org/?domain=github.com
- Examples
  - [hsts](https://hstspreload.org/?domain=github.com)
  - [hsts](https://hstspreload.org/?domain=apviz.io)
  - [hsts](https://hstspreload.org/?domain=baidu.com)

Edit: Replace `no` with `unknow`

:link: **Data**

Where can we get the data from?

- Is there a public API?
  - https://hstspreload.org/api/v2/status?domain=github.com
- Does the API requires an API key?
  - No
- Link to the API documentation.
  - https://hstspreload.org/

Edit: Replace `no` with `unknow`

:microphone: **Motivation**

Please explain why this feature should be implemented and how it would be used.

- What is the specific use case?
  - Show HTST approval status.
  - Track approval and regressions in real time. Alert on potential miss configuration.
  - Promote security for a safer Internet

Kind of related to #2912 and https://github.com/mozilla/http-observatory/issues/96

### PR DESCRIPTION
**Pull request by pxgamer**
As per #2913, this PR adds a new badge for the [HSTS preload][hsts] list.

Badges will use the format `/hsts/:uri.svg` (e.g. `/hsts/github.com.svg`)

Closes #2913 

[hsts]: https://hstspreload.org

### CODE REVIEW

**Comment by pxgamer** [comment]
@yvele, would you say that monitoring or analysis is a better fit for this? 🤔

**Comment by yvele** [comment]
I would say analysis 🤔 there already is vulnerabilities scans in this exact section

**Comment by yvele** [comment]
Maybe a brand new security category would be welcome.. with vulnerabilities, hsts, observatory https://github.com/badges/shields/issues/2912

**Comment by paulmelnikow** [comment]
> Maybe a brand new **security** category would be welcome.. with vulnerabilities, hsts, observatory #2912 🤔

I haven't reviewed this yet, though a thought on the category. 90% of our badges are for software projects and software libraries, and websites and deployed applications are pretty different.

Based on a bunch of issues opened recently, I agree that monitoring, analysis, security, etc. looks like it will be a growing category. It seems pretty clear that hsts and mozilla observatory fit more closely with website, uptime robot, and nodeping (#2910) than with things like snyk and codeclimate. I'm open to suggestions on a better name for that group, though. Maybe we expand it the title to **Hosting and monitoring**?

I'd suggest putting this badge in monitoring for the purpose of this PR, and then opening a second PR when we have consensus on how to rework the categories.

File: services/hsts/hsts.service.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      {
+        title: 'HSTS',
+        namedParams: { uri: 'github.com' },
+        staticPreview: this.render({ status: 'preloaded' }),
+        keywords: ['hsts'],
+      },
+    ]
+  }
+
+  async fetch({ uri }) {
```
[the remark this answers is not in the document]
Comment (pxgamer) [reply services/hsts/hsts.service.js]: Yes, that would make more sense :+1:

File: services/hsts/hsts.service.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
+'use strict'
+
+const label = 'hsts'
+const { BaseJsonService } = require('..')
+
+const Joi = require('joi')
+const schema = Joi.object({
+  name: Joi.string().required(),
+  status: Joi.string().required(),
+  bulk: Joi.boolean().required(),
```
[the remark this answers is not in the document]
Comment (pxgamer) [reply services/hsts/hsts.service.js]: Ok, I wasn't sure whether to include it.

File: services/hsts/hsts.service.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
+'use strict'
+
+const label = 'hsts'
+const { BaseJsonService } = require('..')
+
+const Joi = require('joi')
+const schema = Joi.object({
+  name: Joi.string().required(),
```
Comment (yvele) [inline services/hsts/hsts.service.js]: `name` also is not used in the badge (to be removed from unit tests)

File: services/hsts/hsts.service.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    }
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'HSTS',
+        namedParams: { uri: 'github.com' },
+        staticPreview: this.render({ status: 'preloaded' }),
+        keywords: ['hsts'],
```
[the remark this answers is not in the document]
Comment (pxgamer) [reply services/hsts/hsts.service.js]: I'll wait for some feedback and add it. Accidentally clicked the update branch button _sigh_ So I don't want to push too many changes for now. :yum:

[the remark this answers is not in the document]
Comment (calebcartwright) [reply services/hsts/hsts.service.js]: The content of `title` is automatically searchable (case insensitive) so you don't need to include `hsts` here. I don't think we generally include a categorization as a search keyword but not a big deal IMO if you want to add that as a keyword

File: services/hsts/hsts.service.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+const { BaseJsonService } = require('..')
+
+const Joi = require('joi')
+const schema = Joi.object({
+  status: Joi.string().required(),
+}).required()
+
+module.exports = class HSTS extends BaseJsonService {
+  static get category() {
+    return 'monitoring'
```
Comment (calebcartwright) [inline services/hsts/hsts.service.js]: Would `analysis` be a better category fit here? I've still not had a chance to dive into the details of this yet so honestly not sure what the best category would be

File: services/hsts/hsts.tester.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
+'use strict'
+
+const { ServiceTester } = require('..')
+const label = 'hsts'
+
+const t = (module.exports = new ServiceTester({
+  id: 'hsts',
+  title: 'HSTS Preload',
+}))
```
Comment (calebcartwright) [inline services/hsts/hsts.tester.js]: I think you should be able to use the shorthand function here to replace lines 3, and 6-9 
`const t = (module.exports = require('..').createServiceTester())`

File: services/hsts/hsts.tester.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+t.create('gets the status of an invalid uri')
+  .get('/does-not-exist.json')
+  .expectJSON({
+    name: label,
+    value: 'unknown',
+  })
+
+t.create('gets the hsts status of github (mock)')
+  .get('/github.com.json')
```
Comment (calebcartwright) [inline services/hsts/hsts.tester.js]: Thanks for testing all paths! Another thing we should probably do here is validation of the custom colors. If you add a `style` query param with a value of `_shields_test` then the response will also include the applied color. For example:

`.get('github.com.json?style=_shields_test')`

Then below:

```
.expectJSON({
  name: label,
  value: 'preloaded',
  color: 'brightgreen'
})
```

Comment (pxgamer) [reply services/hsts/hsts.tester.js]: Thanks, I will add these shortly.

**Comment by pxgamer** [comment]
I've implemented the requested changes. 👍 And I've left the badge as `preloaded`, `pending` and `no` until we decide on the prefix.

File: services/hsts/hsts.service.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        staticPreview: this.render({ status: 'preloaded' }),
+        keywords: ['security'],
+      },
+    ]
+  }
+
+  async fetch({ domain }) {
+    return this._requestJson({
+      schema,
+      url: `https://hstspreload.org/api/v2/status?domain=${domain}`,
```
Comment (paulmelnikow) [inline services/hsts/hsts.service.js]: Could you use `options: { qs: { domain } }` here instead? That ensures correct query string escaping regardless of what users might send.

File: services/hsts/hsts.service.js [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+}).required()
+
+module.exports = class HSTS extends BaseJsonService {
+  static get category() {
+    return 'monitoring'
+  }
+
+  static get route() {
+    return {
+      base: 'hsts',
```
Comment (paulmelnikow) [inline services/hsts/hsts.service.js]: How about we use either `hsts-preload` or `hsts/preload`? As discussed in #2913, this badge isn't for hsts as much as it is for the preload list.

Comment (pxgamer) [reply services/hsts/hsts.service.js]: I prefer `hsts/preload` :+1:

File: services/hsts/hsts.service.js (hunk 8 above)
Comment (pxgamer) [reply services/hsts/hsts.service.js]: Sure. I didn't know that was available, would look neater.

**Comment by paulmelnikow** [comment]
Just checked this out in the review app and it looks good. Thanks for all your work on this!

**Comment by pxgamer** [comment]
No problem. I think I'll probably try and update some of the legacy apps for future PRs too. 👍

**Pull request merged**